homeassistant/components/media_player/mediaroom.py

```diff
@@ -9,134 +9,170 @@
+        hass.data[DATA_MEDIAROOM].append(notify.ip_address)
+        new_stb = MediaroomDevice(
+            host=notify.ip_address, device_id=notify.device_uuid,
+            optimistic=False
+        )
+        async_add_devices([new_stb])
 
-    try:
-        for host in hosts:
-            stbs.append(MediaroomDevice(
-                config.get(CONF_NAME),
-                host,
-                config.get(CONF_OPTIMISTIC),
-                config.get(CONF_TIMEOUT)
-            ))
+    if not config.get(CONF_OPTIMISTIC):
+        from pymediaroom import install_mediaroom_protocol
 
-    except ConnectionRefusedError:
-        hass.components.persistent_notification.create(
-            'Error: Unable to initialize mediaroom at {}<br />'
-            'Check its network connection or consider '
-            'using auto discovery.<br />'
-            'You will need to restart hass after fixing.'
-            ''.format(host),
-            title=NOTIFICATION_TITLE,
-            notification_id=NOTIFICATION_ID)
-
-    add_devices(stbs)
+        await install_mediaroom_protocol(responses_callback=callback_notify)
+        _LOGGER.debug("Auto discovery installed")
 
 
 class MediaroomDevice(MediaPlayerDevice):
     """Representation of a Mediaroom set-up-box on the network."""
 
-    def __init__(self, name, host, optimistic=False, timeout=DEFAULT_TIMEOUT):
+    def set_state(self, mediaroom_state):
+        """Helper method to map pymediaroom states to HA states."""
+        from pymediaroom import State
+
+        state_map = {
+            State.OFF: STATE_OFF,
+            State.STANDBY: STATE_STANDBY,
+            State.PLAYING_LIVE_TV: STATE_PLAYING,
+            State.PLAYING_RECORDED_TV: STATE_PLAYING,
+            State.PLAYING_TIMESHIFT_TV: STATE_PLAYING,
+            State.STOPPED: STATE_PAUSED,
+            State.UNKNOWN: STATE_UNAVAILABLE
+            }
+
+        self._state = state_map[mediaroom_state]
+
+    def __init__(self, host, device_id, optimistic=False,
+                 timeout=DEFAULT_TIMEOUT):
         """Initialize the device."""
         from pymediaroom import Remote
 
-        self.stb = Remote(host, timeout=timeout)
-        _LOGGER.info(
-            "Found %s at %s%s", name, host,
-            " - I'm optimistic" if optimistic else "")
-        self._name = name
-        self._is_standby = not optimistic
-        self._current = None
+        self.host = host
+        self.stb = Remote(host)
+        _LOGGER.info("Found STB at %s%s", host,
+                     " - I'm optimistic" if optimistic else "")
+        self._channel = None
         self._optimistic = optimistic
-        self._state = STATE_STANDBY
+        self._state = STATE_PLAYING if optimistic else STATE_STANDBY
+        self._name = '{} {}'.format(PLATFORM, device_id)
+        if device_id:
+            self._unique_id = device_id
+        else:
+            self._unique_id = self.stb.device_id
```

Comment:
Why wouldn't you always make it `self.stb.device_id` ?

Reply:
This has to do with the opportunistic mode:
- A normal household setup (only one network) will get discovery and the device_id provided is a UUID provided by the STB
- A household where there are more then one network and in which HA and the STB are in different networks will not receive the multicasted message that is used in discovery and state update. In this case I have no access to the UUID and I need to generate a pseudo device_id (self.stb.device_id, which happens to be based on IP address). 

Please note that the later is a completely non standard setup and not supported by the IPTV operators using Mediaroom, so we are talking about expert users who have setup their own home network infrastructure. 

I see no other way around this, unless to make unique_id optional or dropping support for the optimistic mode.

Reply:
Okay, so we should never have a unique ID based on IP address. It is totally fine to return only the unique ID when you have access to it. Return `None` otherwise.